### PR TITLE
Added GraphQL query

### PR DESCRIPTION
**Pull request by mpastorek**
- added new function to BlueprintRepository that is responsible for reading blueprints from DB based on defined filters on the input.
- added BlueprintReader interface and its implementation that uses function to read blueprints from BlueprintRepository.
- added new GraphQL models: BlueprintConnection, BlueprintEdge and PageInfo.
- fixed ordering of the enum connection_type in the V1 migration. Now it is still safe in this way.
- added app.props to application.yml to print SQL queries in the log
- updated port of the testing Postgresql db to better testing
- added tests
- updated README (ktlint)

### CODE REVIEW

**Review by MatejGlemba — APPROVED** [review]

File: src/main/kotlin/graphql/queries/BlueprintQueries.kt [hunk 1]
```diff
@@ -23,13 +23,74 @@
...
+    fun blueprints(
+        @GraphQLDescription("The management protocol used for creation of connection to the device.")
+        connectionType: ConnectionType? = null,
+        @GraphQLDescription("The vendor of the device (for example, 'cisco').")
+        vendor: String? = null,
+        @GraphQLDescription("The model of the device (for example, 'ios xr').")
+        model: String? = null,
+        @GraphQLDescription("The version of the device (for example, '6').")
+        version: String? = null,
+        limit: Int? = null,
```
Comment (jaro0149) [inline src/main/kotlin/graphql/queries/BlueprintQueries.kt:86]: please, use standard naming: https://relay.dev/graphql/connections.htm
limit -> first
blueprintNode -> node

File: src/main/kotlin/graphql/queries/BlueprintQueries.kt [hunk 2]
```diff
@@ -23,13 +23,74 @@
...
+        fun List<BlueprintEdge>.getPageInfo(hasNext: Boolean) = PageInfo(hasNext, this.getCursor())
+
+        private fun List<BlueprintEdge>.getCursor(): String? = if (this.isNotEmpty()) this.last().cursor else null
+
+        private fun List<BlueprintData>.limitIfHasNext(
+            hasNext: Boolean,
+            limit: Int?,
+        ): List<BlueprintData> = if (limit != null && hasNext) this.take(limit) else this
+
+        private fun BlueprintData.toBlueprintEdge(): BlueprintEdge =
```
Comment (jaro0149) [inline src/main/kotlin/graphql/queries/BlueprintQueries.kt:63]: there is already GraphQLConverters file that contains such converters

File: src/main/kotlin/graphql/model/BlueprintEdge.kt [hunk 3]
```diff
@@ -0,0 +1,34 @@
...
+ * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+ * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
+ * SOFTWARE.
+ */
+package com.elisapolystar.blueprints.graphql.model
+
+import com.expediagroup.graphql.generator.annotations.GraphQLDescription
+
+@GraphQLDescription("Represents an edge in the blueprint graph.")
+data class BlueprintEdge(
```
Comment (jaro0149) [inline src/main/kotlin/graphql/model/BlueprintEdge.kt:29]: BluperintEdge and BluperintConnection classes can be made generic -> Edge<BluperintNode> and Connection<Edge<BlueprintNode>>

File: src/main/kotlin/services/model/BlueprintData.kt [hunk 4]
```diff
@@ -30,7 +30,7 @@ import com.google.gson.annotations.JsonAdapter
 /**
  * Information about the blueprint - identifier, patterns used for matching the device, and the content of the template.
  */
-internal data class BlueprintData(
```
Comment (jaro0149) [inline src/main/kotlin/services/model/BlueprintData.kt:33]: why? it is internal service that should be used only in this module
only models in the graphql.model package should be exposed

File: src/main/kotlin/services/model/PersistedBlueprint.kt [hunk 5]
```diff
@@ -28,7 +28,7 @@ import java.time.LocalDateTime
 /**
  * Blueprint that is persisted in the datastore.
  */
-internal data class PersistedBlueprint(
```
Comment (jaro0149) [inline src/main/kotlin/services/model/PersistedBlueprint.kt:31]: why?

File: src/main/kotlin/services/api/BlueprintWriter.kt [hunk 6]
```diff
@@ -29,7 +29,7 @@ import com.elisapolystar.blueprints.services.model.PersistedBlueprint
 /**
  * Services used for creation, removing, and updating of persisted blueprints.
  */
-internal interface BlueprintWriter {
+interface BlueprintWriter {
```
Comment (jaro0149) [inline src/main/kotlin/services/api/BlueprintWriter.kt:32]: internal interface should be sufficient, we are not using this service outside of this module

File: src/main/kotlin/services/api/BlueprintReader.kt [hunk 7]
```diff
@@ -0,0 +1,53 @@
...
+ */
+package com.elisapolystar.blueprints.services.api
+
+import com.elisapolystar.blueprints.commons.ConnectionType
+import com.elisapolystar.blueprints.services.model.PersistedBlueprint
+
+/**
+ * Service responsible for reading persisted blueprint data from the datastore.
+ */
+interface BlueprintReader {
```
Comment (jaro0149) [inline src/main/kotlin/services/api/BlueprintReader.kt:32]: internal service

File: src/main/kotlin/graphql/queries/BlueprintQueries.kt [hunk 8]
```diff
@@ -23,13 +23,74 @@
...
+                .toList()
+
+        fun List<BlueprintEdge>.getPageInfo(hasNext: Boolean) = PageInfo(hasNext, this.getCursor())
+
+        private fun List<BlueprintEdge>.getCursor(): String? = if (this.isNotEmpty()) this.last().cursor else null
+
+        private fun List<BlueprintData>.limitIfHasNext(
+            hasNext: Boolean,
+            limit: Int?,
+        ): List<BlueprintData> = if (limit != null && hasNext) this.take(limit) else this
```
Comment (jaro0149) [inline src/main/kotlin/graphql/queries/BlueprintQueries.kt:61]: why is this necessary? limiting should be done on DB side

File: src/main/kotlin/graphql/queries/BlueprintQueries.kt [hunk 9]
```diff
@@ -23,13 +23,74 @@
...
+
+        private fun BlueprintData.toBlueprintEdge(): BlueprintEdge =
+            BlueprintEdge(
+                this.name,
```
Comment (jaro0149) [inline src/main/kotlin/graphql/queries/BlueprintQueries.kt:65]: please, used named arguments, it is more readable

Comment (jaro0149) [inline src/main/kotlin/graphql/queries/BlueprintQueries.kt:65]: I think, it is more common to use something that is opaque to user as cursor, for example blueprint/1 (id)

File: src/main/kotlin/database/jpa/BlueprintEntity.kt [hunk 10]
```diff
@@ -52,7 +52,7 @@ import java.time.ZoneOffset
         ),
     ],
 )
-internal class BlueprintEntity(
+class BlueprintEntity(
```
Comment (jaro0149) [inline src/main/kotlin/database/jpa/BlueprintEntity.kt:55]: database layer should be completely hidden in the module

File: src/main/kotlin/services/blueprints/BlueprintDatabaseReader.kt [hunk 11]
```diff
@@ -0,0 +1,65 @@
...
+            connectionType?.name ?: "",
+            vendorPattern ?: "",
+            modelPattern ?: "",
+            versionPattern ?: "",
+        )
+        val blueprints =
+            blueprintRepository
+                .findAllByPatterns(connectionType?.name, vendorPattern, modelPattern, versionPattern, limit, after)
+                .map { entity -> entity.toPersistedData() }
+                .toList()
```
Comment (jaro0149) [inline src/main/kotlin/services/blueprints/BlueprintDatabaseReader.kt:61]: one additional thing that would be good to add is ordering of found blueprints by 'the best match' algorithm
- the first returned blueprint should be one that contains the most of the specified parameters from the input
- the last returned blueprint should be a generic blueprint, for example, for all CLI devices without specified model/version/vendor

File: src/main/kotlin/services/model/BlueprintData.kt (hunk 4 above)
Comment (mpastorek) [reply src/main/kotlin/services/model/BlueprintData.kt:33]: BlueprintWriter and BlueprintReader uses this model/entity classes. These two servuces should be public because mutations/queries use them. I changed DI of these services in the BlueprintMutations and BlueprintQueries from field injection to constructor injection, that is preffered. At that moment these services and also entity classes cannot be internal.

File: src/main/kotlin/services/api/BlueprintWriter.kt (hunk 6 above)
Comment (mpastorek) [reply src/main/kotlin/services/api/BlueprintWriter.kt:32]: we use constructor injection of this service. it cannot be internal

File: src/main/kotlin/services/api/BlueprintReader.kt (hunk 7 above)
Comment (mpastorek) [reply src/main/kotlin/services/api/BlueprintReader.kt:32]: again constructor injection

File: src/main/kotlin/graphql/queries/BlueprintQueries.kt (hunk 1 above)
Comment (mpastorek) [reply src/main/kotlin/graphql/queries/BlueprintQueries.kt:86]: done

File: src/main/kotlin/graphql/queries/BlueprintQueries.kt (hunk 2 above)
Comment (mpastorek) [reply src/main/kotlin/graphql/queries/BlueprintQueries.kt:63]: done

File: src/main/kotlin/graphql/model/BlueprintEdge.kt (hunk 3 above)
Comment (mpastorek) [reply src/main/kotlin/graphql/model/BlueprintEdge.kt:29]: done

File: src/main/kotlin/graphql/queries/BlueprintQueries.kt (hunk 8 above)
Comment (mpastorek) [reply src/main/kotlin/graphql/queries/BlueprintQueries.kt:61]: done

File: src/main/kotlin/graphql/queries/BlueprintQueries.kt (hunk 9 above)
Comment (mpastorek) [reply src/main/kotlin/graphql/queries/BlueprintQueries.kt:65]: this method has been removed

Comment (mpastorek) [reply src/main/kotlin/graphql/queries/BlueprintQueries.kt:65]: cursor has to be unique. only name in the blueprint table is unique. what do you mean with blueprint/1? Database record PK + name?

Comment (jaro0149) [reply src/main/kotlin/graphql/queries/BlueprintQueries.kt]: '1' in this example value of 'id' field in database that is unique (primary key)

Comment (mpastorek) [reply src/main/kotlin/graphql/queries/BlueprintQueries.kt]: yes, I fixed it

File: src/main/kotlin/services/blueprints/BlueprintDatabaseReader.kt (hunk 11 above)
Comment (mpastorek) [reply src/main/kotlin/services/blueprints/BlueprintDatabaseReader.kt:84]: done

**Review by jaro0149 — APPROVED** [review]
